src/editor/write_level.cpp

```diff
@@ -0,0 +1,406 @@
+        section.wrap_v = LevelVWrap[i];
+        section.OffScreenEn = OffScreenExit[i];
+        section.background = Background2[i];
+        section.lock_left_scroll = NoTurnBack[i];
+        section.underwater = UnderWater[i];
+        section.music_file = CustomMusic[i];
+
+        // swapped order of operands because
+        // previous code did not make sense
+        if(size_t(i) >= out.sections.size())
+            out.sections.push_back(section);
+        else
+            out.sections[i] = section;
+    }
+
+    for(int i = 1; i <= 2; ++i)
+    {
+        auto &p = PlayerStart[i];
+        player.id = i;
+        player.x = p.X;
+        player.y = p.Y;
+        player.w = p.Width;
+        player.h = p.Height;
+        player.direction = p.Direction;
+        out.players.push_back(player);
+    }
+
+    for(int i = 1; i <= numBlock; ++i)
+    {
+        auto &b = Block[i];
+
+        block.id = b.Type;
+        block.x = b.Location.X;
+        block.y = b.Location.Y;
+        block.w = b.Location.Width;
+        block.h = b.Location.Height;
+
+        if(b.Special >= 1000)
+            block.npc_id = b.Special - 1000;
+        else if(b.Special > 0)
+            block.npc_id = -b.Special;
+        else
+            block.npc_id = 0;
+
+        block.invisible = b.Invis;
+        block.slippery = b.Slippy;
+        block.layer = GetL(b.Layer);
+
+        // fix this to update as needed
+        if(block.layer.empty())
+            block.layer = "Default";
+
+        block.event_destroy = GetE(b.TriggerDeath);
+        block.event_hit = GetE(b.TriggerHit);
+        block.event_emptylayer = GetE(b.TriggerLast);
+
+        block.meta.array_id = (out.blocks_array_id++);
+
+        out.blocks.push_back(block);
+    }
+
+    for(int i = 1; i <= numBackground; ++i)
+    {
+        auto &b = Background[i];
+
+        bgo.id = b.Type;
+        bgo.x = b.Location.X;
+        bgo.y = b.Location.Y;
+        bgo.layer = GetL(b.Layer);
+
+        bgo.z_mode = b.zMode;
+        bgo.z_offset = b.zOffset;
+        bgo.smbx64_sp = bgo.z_mode == LevelBGO::ZDefault ? b.SortPriority : -1;
+
+        // fix this to update as needed
+        if(bgo.layer.empty())
+            bgo.layer = "Default";
+
+        bgo.meta.array_id = out.bgo_array_id++;
+        out.bgo.push_back(bgo);
+    }
+
+    for(int i = 1; i <= numNPCs; ++i)
+    {
+        auto &n = NPC[i];
+
+        npc.id = n.Type;
+        npc.x = n.Location.X;
+        npc.y = n.Location.Y;
+        npc.direct = n.Direction;
+
+        if(n.Type == 91 || n.Type == 96 || n.Type == 283 || n.Type == 284)
+            npc.contents = n.Special;
+
+        if(n.Type == 288 || n.Type == 289 || (n.Type == 91 && n.Special == 288))
+            npc.special_data = n.Special2;
+        else if(n.Type == 260 || NPCIsAParaTroopa[n.Type] || NPCIsCheep[n.Type])
+            npc.special_data = n.Special;
+        else if(n.Type == 86)
+            npc.special_data = n.Special7;
```

Comment:
Since the time, there are more NPCs got the support for Special7 such as rail platforms, bill blasters, etc.

Reply:
Should we just set `npc.special_data = n.Special7` for *everything*?

Reply:
:thinking: 
Not so, just for these NPC that got this rule at the side of the main level file loader. There are some custom rules that could be done.